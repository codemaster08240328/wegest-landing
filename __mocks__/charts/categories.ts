export const categories = [
  {
    id: 'Service',
    value: 93,
    color: '#ff4567',
  },
  {
    id: 'Product',
    value: 38,
    color: '#99AAFF',
  },
]
