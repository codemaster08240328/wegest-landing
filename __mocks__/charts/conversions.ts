export const conversions = [
  {
    id: 'new_man',
    label: 'New Man',
    value: 25,
    color: 'hsl(172, 70%, 50%)',
  },
  {
    id: 'habitual_man',
    label: 'Habitual Man',
    value: 123,
    color: 'hsl(245, 70%, 50%)',
  },
  {
    id: 'new_woman',
    label: 'New Woman',
    value: 58,
    color: 'hsl(72, 70%, 50%)',
  },
  {
    id: 'habitual_woman',
    label: 'Habitual Woman',
    value: 154,
    color: 'hsl(33, 70%, 50%)',
  },
]
