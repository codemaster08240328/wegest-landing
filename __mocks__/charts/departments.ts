export const departments = [
  {
    department: 'Department 1',
    value: 175,
    valueColor: '#81C884',
  },
  {
    department: 'Department 2',
    value: 235,
    valueColor: '#81C884',
  },
  {
    department: 'Department 3',
    value: 120,
    valueColor: '#81C884',
  },
  {
    department: 'Department 4',
    value: 298,
    valueColor: '#81C884',
  },
  {
    department: 'Department 5',
    value: 345,
    valueColor: '#81C884',
  },
  {
    department: 'Department 6',
    value: 200,
    valueColor: '#81C884',
  },
  {
    department: 'Department 7',
    value: 380,
    valueColor: '#81C884',
  },
]
