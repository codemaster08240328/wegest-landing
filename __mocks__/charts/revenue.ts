interface IMonth {
  month: string
  2019: number
  2020: number
  2021: number
  '2019Color': string
  '2020Color': string
  '2021Color': string
}

interface IDepartment {
  department: string
  value: number
  valueColor: string
}

const dataColors = {
  '2019Color': '#81C884',
  '2020Color': '#E37372',
  '2021Color': '#64B5F6',
}

export const revenue: IMonth[] = [
  {
    month: 'Jan',
    2019: 12175,
    2020: 5503,
    2021: 3151,
    ...dataColors,
  },
  {
    month: 'Feb',
    2019: 20778,
    2020: 2846,
    2021: 3446,
    ...dataColors,
  },
  {
    month: 'Mar',
    2019: 10838,
    2020: 5836,
    2021: 3446,
    ...dataColors,
  },
  {
    month: 'Apr',
    2019: 4730,
    2020: 1915,
    2021: 4676,
    ...dataColors,
  },
  {
    month: 'May',
    2019: 3317,
    2020: 2363,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Jun',
    2019: 11273,
    2020: 2570,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Jul',
    2019: 18117,
    2020: 2425,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Aug',
    2019: 4658,
    2020: 2844,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Sep',
    2019: 3943,
    2020: 2108,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Oct',
    2019: 1174,
    2020: 4013,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Nov',
    2019: 9833,
    2020: 423,
    2021: 0,
    ...dataColors,
  },
  {
    month: 'Dec',
    2019: 1524,
    2020: 4697,
    2021: 0,
    ...dataColors,
  },
]

const getMonthValue = (index: number, year: string) => {
  const value: number = (revenue[index] as any)[year]

  if (Number(value) === 0) return 0

  return (value / 1000).toFixed(3)
}

const getYearTotal = (year: string) => {
  return (
    revenue.reduce((sum, month) => Number((month as any)[year]) + sum, 0) / 1000
  ).toFixed(3)
}

export const revenueTable = [
  {
    year: '2019',
    Jan: getMonthValue(0, '2019'),
    Feb: getMonthValue(1, '2019'),
    Mar: getMonthValue(2, '2019'),
    Apr: getMonthValue(3, '2019'),
    May: getMonthValue(4, '2019'),
    Jun: getMonthValue(5, '2019'),
    Jul: getMonthValue(6, '2019'),
    Aug: getMonthValue(7, '2019'),
    Sep: getMonthValue(8, '2019'),
    Oct: getMonthValue(9, '2019'),
    Nov: getMonthValue(10, '2019'),
    Dec: getMonthValue(11, '2019'),
    Total: getYearTotal('2019'),
  },
  {
    year: '2020',
    Jan: getMonthValue(0, '2020'),
    Feb: getMonthValue(1, '2020'),
    Mar: getMonthValue(2, '2020'),
    Apr: getMonthValue(3, '2020'),
    May: getMonthValue(4, '2020'),
    Jun: getMonthValue(5, '2020'),
    Jul: getMonthValue(6, '2020'),
    Aug: getMonthValue(7, '2020'),
    Sep: getMonthValue(8, '2020'),
    Oct: getMonthValue(9, '2020'),
    Nov: getMonthValue(10, '2020'),
    Dec: getMonthValue(11, '2020'),
    Total: getYearTotal('2020'),
  },
  {
    year: '2021',
    Jan: getMonthValue(0, '2021'),
    Feb: getMonthValue(1, '2021'),
    Mar: getMonthValue(2, '2021'),
    Apr: getMonthValue(3, '2021'),
    May: getMonthValue(4, '2021'),
    Jun: getMonthValue(5, '2021'),
    Jul: getMonthValue(6, '2021'),
    Aug: getMonthValue(7, '2021'),
    Sep: getMonthValue(8, '2021'),
    Oct: getMonthValue(9, '2021'),
    Nov: getMonthValue(10, '2021'),
    Dec: getMonthValue(11, '2021'),
    Total: getYearTotal('2021'),
  },
]
