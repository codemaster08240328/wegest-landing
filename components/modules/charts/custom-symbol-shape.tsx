interface CustomSymbolShapeProps {
  x: number
  y: number
  size: number
  fill: string
  borderWidth: number
  borderColor: string
}

const CustomSymbolShape = ({
  x,
  y,
  size,
  fill,
  borderWidth,
  borderColor,
}: CustomSymbolShapeProps) => {
  const hidden = fill === 'transparent'

  return (
    <rect
      x={x}
      y={y}
      fill={hidden ? 'transparent' : fill}
      strokeWidth={hidden ? 2 : borderWidth}
      stroke={hidden ? '#AAA' : borderColor}
      width={size}
      height={size}
      style={{ pointerEvents: 'none' }}
    />
  )
}

export default CustomSymbolShape
