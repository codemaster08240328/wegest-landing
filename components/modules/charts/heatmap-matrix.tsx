import React from 'react'
import { AxisProps } from '@nivo/axes'
import { ResponsiveHeatMapCanvas, HeatMapCanvasProps } from '@nivo/heatmap'

interface MyResponsiveHeatMapProps extends HeatMapCanvasProps {}

type Axis = AxisProps & {
  renderTick?: null
  orient?: 'top' | 'right' | 'bottom' | 'left'
}

const defaultLeftAxis: Axis = {
  orient: 'left',
  tickSize: 0,
  tickPadding: 5,
  tickRotation: 0,
  legend: '',
  legendPosition: 'middle',
  legendOffset: 0,
}

const defaultTopAxis: Axis = {
  orient: 'top',
  tickSize: 0,
  tickPadding: 5,
  tickRotation: -70,
  legend: '',
  legendOffset: 36,
}

const MyResponsiveHeatMap = ({
  data,
  keys,
  indexBy,
  margin = { top: 20, right: 0, bottom: 0, left: 30 },
  padding = 2,
  axisTop = defaultTopAxis,
  axisLeft = defaultLeftAxis,
}: MyResponsiveHeatMapProps) => (
  <div style={{ width: '100%', height: 120 }}>
    <ResponsiveHeatMapCanvas
      forceSquare
      data={data}
      keys={keys}
      indexBy={indexBy}
      margin={margin}
      padding={padding}
      axisTop={axisTop}
      axisLeft={axisLeft}
      cellOpacity={1}
      labelTextColor='transparent'
      hoverTarget='cell'
      cellHoverOthersOpacity={0.6}
    />
  </div>
)

export default MyResponsiveHeatMap
