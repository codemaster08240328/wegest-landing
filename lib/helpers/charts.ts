export function toggleLegendOnClick(data: any, setData: (d: any) => void) {
  setData((previousData: any[]) => {
    return previousData.map((item: any) => {
      const isItem = item.id === data.id
      const newItem = { ...item }
      const hide = isItem ? !item.hide : item.hide

      if (item.hide || previousData.filter((d) => !d.hide).length > 1) {
        newItem.hide = hide

        if (hide && isItem) {
          newItem.originalValue = newItem.value
          newItem.originalColor = newItem.color
          newItem.value = 0
          newItem.color = 'transparent'
        } else if (!hide && isItem) {
          if (newItem.originalValue) newItem.value = newItem.originalValue
          if (newItem.originalColor) newItem.color = newItem.originalColor

          delete newItem.originalValue
          delete newItem.originalColor
        }
      }

      return newItem
    })
  })
}
