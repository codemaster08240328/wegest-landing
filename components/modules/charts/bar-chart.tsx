import React, { useState } from 'react'
import { AxisProps } from '@nivo/axes'
import { LegendProps, LegendAnchor } from '@nivo/legends'
import {
  ResponsiveBar,
  ResponsiveBarCanvas,
  BarSvgProps,
  BarCanvasProps,
} from '@nivo/bar'

import { toggleLegendOnClick } from '@/lib/helpers'
import CustomSymbolShape from './custom-symbol-shape'

type DataFrom = 'indexes' | 'keys'
type BarProps = BarSvgProps & BarCanvasProps
interface MyResponsiveBarProps extends BarProps {
  type?: 'svg' | 'canvas'
  legendIsClickable?: boolean
  legendAnchor?: LegendAnchor
  legendDataFrom?: DataFrom
}

type Axis = AxisProps & {
  renderTick?: null
}

const defaultLeftAxis: Axis = {
  tickSize: 5,
  tickPadding: 5,
  tickRotation: -60,
  legend: '',
  legendPosition: 'middle',
  legendOffset: -40,
}

const defaultBottomAxis: Axis = {
  tickSize: 5,
  tickPadding: 5,
  tickRotation: 0,
  legend: '',
  legendPosition: 'middle',
  legendOffset: 32,
}

const defaultLegends: Array<{ dataFrom: DataFrom } & LegendProps> = [
  {
    dataFrom: 'keys',
    anchor: 'bottom',
    direction: 'row',
    justify: false,
    translateX: 0,
    translateY: 55,
    itemsSpacing: 10,
    itemWidth: 65,
    itemHeight: 20,
    itemDirection: 'left-to-right',
    itemOpacity: 0.85,
    symbolSize: 15,
    effects: [
      {
        on: 'hover',
        style: {
          itemOpacity: 1,
        },
      },
    ],
  },
]

const BarChart = ({ type, ...props }: MyResponsiveBarProps) => {
  if (type === 'svg') return <ResponsiveBar {...props} />

  return <ResponsiveBarCanvas {...props} />
}

const MyResponsiveBar = ({
  data: initialData,
  keys,
  type,
  indexBy,
  layout = 'vertical',
  colors = { scheme: 'nivo' },
  margin = { top: 10, right: 10, bottom: 70, left: 30 },
  padding = 0.2,
  innerPadding = 3,
  groupMode = 'grouped',
  axisLeft = defaultLeftAxis,
  axisBottom = defaultBottomAxis,
  legends = defaultLegends,
  legendIsClickable = false,
  legendAnchor,
  legendDataFrom,
}: MyResponsiveBarProps) => {
  const [data, setData] = useState(initialData)

  if (legends.length > 0) {
    if (legendIsClickable) {
      legends[0].symbolShape = CustomSymbolShape
      legends[0].onClick = (d) => toggleLegendOnClick(d, setData)
    }

    if (legendAnchor) legends[0].anchor = legendAnchor
    if (legendDataFrom) legends[0].dataFrom = legendDataFrom
  }

  return (
    <BarChart
      enableGridX
      data={data}
      minValue={0}
      type={type}
      layout={layout}
      keys={keys}
      indexBy={indexBy}
      margin={margin}
      padding={padding}
      innerPadding={innerPadding}
      groupMode={groupMode}
      valueScale={{ type: 'linear' }}
      indexScale={{ type: 'band', round: true }}
      colors={colors}
      borderColor={{ from: 'color', modifiers: [['darker', 1.6]] }}
      axisTop={null}
      axisRight={null}
      axisBottom={axisBottom}
      axisLeft={axisLeft}
      labelSkipWidth={12}
      labelSkipHeight={12}
      labelTextColor={{ from: 'color', modifiers: [['darker', 1.6]] }}
      legends={legends}
    />
  )
}

export default MyResponsiveBar
