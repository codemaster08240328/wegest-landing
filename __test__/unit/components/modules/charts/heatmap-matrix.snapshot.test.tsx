import renderer from 'react-test-renderer'

import HeatmapGraph from '@/components/modules/charts/heatmap-matrix'
import { revenueTable } from '@/__mocks__/charts/revenue'

describe('<HeatmapGraph> snapshot', () => {
  it('renders correctly', () => {
    expect(
      renderer
        .create(
          <HeatmapGraph
            data={revenueTable}
            indexBy='year'
            keys={[
              'Jan',
              'Feb',
              'Mar',
              'Apr',
              'May',
              'Jun',
              'Jul',
              'Aug',
              'Sep',
              'Oct',
              'Nov',
              'Dec',
              'Total',
            ]}
          />,
        )
        .toJSON(),
    ).toMatchSnapshot()
  })
})
