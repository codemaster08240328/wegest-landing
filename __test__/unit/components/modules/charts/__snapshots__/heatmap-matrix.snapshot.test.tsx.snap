// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeatmapGraph> snapshot renders correctly 1`] = `
<div
  style={
    Object {
      "height": 120,
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
</div>
`;
